### PR TITLE
Add `second_factors`

### PR DESCRIPTION
**Pull request by Joerger**
Add `second_factors` and prefer it over `second_factor`.

We don't currently plan on removing second_factor, as this would require a more complicated migration process. Instead we will just derive `second_factors` from `second_factor` (and vice versa) and output a warning log when `second_factor` is set.

There is no plan to deprecate `second_factor` completely. When `second_factor` is set and `second_factors` is not, or vice versa, we convert from one to the other.

In a follow up PR I will update as much logic as possible to use `second_factors` instead of `second_factor`, as they are two sources of the same information.

In this PR I've also added the SSO second_factor type. It is currently completely unused, but we'd rather get the proto changes into v17 rather than waiting until SSO MFA is fully released in a minor version.

Follow up TODO: Update docs.

Changelog: Add new second_factors field to cluster auth preference for more clarity and granularity over which 2fa methods are enabled in a cluster.

Depends on https://github.com/gravitational/teleport/pull/47230

### CODE REVIEW

File: api/types/authentication.go [hunk 1]
```diff
@@ -74,12 +75,17 @@ type AuthPreference interface {
...
+	// SetSecondFactors sets the list of supported second factors.
+	SetSecondFactors([]SecondFactorType)
 	// GetPreferredLocalMFA returns a server-side hint for clients to pick an MFA
 	// method when various options are available.
 	// It is empty if there is nothing to suggest.
 	GetPreferredLocalMFA() constants.SecondFactorType
-	// IsSecondFactorEnforced checks if second factor is enforced
-	// (not disabled or set to optional).
+	// IsSecondFactorEnabled checks if second factor is enabled.
+	IsSecondFactorEnabled() bool
```
Comment (Joerger) [inline api/types/authentication.go:87]: note: this will be used in the follow up PR, but I'd rather add it here with these changes.

File: api/proto/teleport/legacy/types/types.proto [hunk 2]
```diff
@@ -2099,6 +2099,20 @@ message AuthPreferenceSpecV2 {
   // SignatureAlgorithmSuite is the configured signature algorithm suite for the cluster.
   // The current default value is "legacy". This field is not yet fully supported.
   SignatureAlgorithmSuite signature_algorithm_suite = 20;
+
+  // SecondFactors is a list of supported second factor types.
+  repeated SecondFactorType SecondFactors = 21 [(gogoproto.jsontag) = "second_factors,omitempty"];
```
Comment (tigrato) [inline api/proto/teleport/legacy/types/types.proto:2104]:
```suggestion
  repeated SecondFactorType allowed_second_factors = 21 [(gogoproto.jsontag) = "supported_second_factors,omitempty"];
```

File: api/types/authentication.go [hunk 3]
```diff
@@ -314,62 +320,119 @@ func (c *AuthPreferenceV2) SetType(s string) {
...
 }
 
 // SetSecondFactor sets the type of second factor.
 func (c *AuthPreferenceV2) SetSecondFactor(s constants.SecondFactorType) {
 	c.Spec.SecondFactor = s
+
+	// SecondFactors takes precedence, so if SecondFactor is being set we should
+	// unsetSecondFactors here. SecondFactors will still be derived from SecondFactor
+	// configuration and set by CheckAndSetDefaults.
+	c.Spec.SecondFactors = nil
```
Comment (tigrato) [inline api/types/authentication.go:355]: if our goal is to migrate away from string, should we convert the second factor to SecondFactors?

File: api/types/authentication.go [hunk 4]
```diff
@@ -314,62 +320,119 @@ func (c *AuthPreferenceV2) SetType(s string) {
...
 	default:
 		slog.WarnContext(context.Background(), "Found unknown second_factor setting", "second_factor", sf)
-		return "" // Unsure, say nothing.
+		return nil
+	}
+}
+
+// SetSecondFactors sets the list of supported second factors.
+func (c *AuthPreferenceV2) SetSecondFactors(s []SecondFactorType) {
+	c.Spec.SecondFactors = s
```
Comment (tigrato) [inline api/types/authentication.go:393]: since SecondFactor takes precedence, should we also handle it here?

File: api/proto/teleport/legacy/types/types.proto (hunk 2 above)
Comment (Joerger) [reply api/proto/teleport/legacy/types/types.proto:2104]: @rosstimothy WDYT about `supported_second_factors`? Product already approved `second_factors` right?

File: lib/config/fileconf.go [hunk 5]
```diff
@@ -1084,9 +1085,28 @@ func (a *AuthenticationConfig) Parse() (types.AuthPreference, error) {
...
+			log.Warnf(`` +
+				`second_factor and second_factors are both set. second_factors will take precedence. ` +
+				`second_factor should be unset to remove this warning.`)
+		}
+	}
+
 	return types.NewAuthPreferenceFromConfigFile(types.AuthPreferenceSpecV2{
 		Type:                    a.Type,
 		SecondFactor:            a.SecondFactor,
+		SecondFactors:           a.SecondFactors,
```
Comment (codingllama) [inline lib/config/fileconf.go:1109]: Similar to other comments, won't this cause changes to CAP?

**Review by codingllama — COMMENTED** [review]
Thanks for the separate PR, Brian! Apologies for the delay.

File: api/types/authentication.go [hunk 6]
```diff
@@ -685,59 +743,38 @@ func (c *AuthPreferenceV2) CheckAndSetDefaults() error {
...
-			if err := c.Spec.U2F.Check(); err != nil {
-				return trace.Wrap(err)
-			}
-			if c.Spec.Webauthn == nil {
-				// Not a problem, try to derive from U2F.
-				c.Spec.Webauthn = &Webauthn{}
-			}
+	// If U2F is present validate it, we can derive Webauthn from it.
+	if c.Spec.U2F != nil {
+		if err := c.Spec.U2F.Check(); err != nil {
```
[the remark this answers is not in the document]
Comment (codingllama) [reply api/types/authentication.go:748]: I think the comment still applies for second_factor="otp".

File: lib/config/fileconf.go [hunk 7]
```diff
@@ -1084,9 +1085,28 @@ func (a *AuthenticationConfig) Parse() (types.AuthPreference, error) {
...
+	if a.SecondFactor != "" {
+		secondFactors := types.SecondFactorsFromLegacySecondFactor(a.SecondFactor)
+		var secondFactorStrings []string
+		for _, sf := range secondFactors {
+			secondFactorStrings = append(secondFactorStrings, sf.ToString())
+		}
+
+		log.Warnf(``+
+			`The "second_factor" setting is marked for removal in favor of second_factors. `+
+			`Please update your configuration to use second_factors. e.g. "second_factors: %v".`, secondFactorStrings)
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/config/fileconf.go:1097]: even so I'm of the option it should stay, but I'll yield to UX as usual.

File: api/types/authentication.go [hunk 8]
```diff
@@ -314,62 +320,117 @@ func (c *AuthPreferenceV2) SetType(s string) {
...
+
+	// Set SecondFactor from SecondFactors.
+	c.Spec.SecondFactor = legacySecondFactorFromSecondFactors(c.Spec.SecondFactors)
+}
+
+// GetPreferredLocalMFA returns a server-side hint for clients to pick an MFA
+// method when various options are available.
+// It is empty if there is nothing to suggest.
+func (c *AuthPreferenceV2) GetPreferredLocalMFA() constants.SecondFactorType {
+	if c.IsSecondFactorWebauthnAllowed() {
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/authentication.go:398]: No, we're just going to prefer WebAuthn > SSO > TOTP rather than an ordered list.

File: lib/config/fileconf.go (hunk 5 above)
Comment (Joerger) [reply lib/config/fileconf.go:1109]: I'm not sure what you mean

Comment (codingllama) [reply lib/config/fileconf.go:1109]: I mean that we are writing both second_factor and second_factors to the "cluster_auth_preference". This means that, on a read, we'll return a resource that has both fields set and thus is invalid by Teleport's own criteria.

This is all to say I would rather keep the cap as the user wrote it and calculate whatever we need in the getters, on the fly.

File: api/types/authentication.go [hunk 9]
```diff
@@ -314,62 +320,117 @@ func (c *AuthPreferenceV2) SetType(s string) {
...
+		return []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_OTP}
+	case constants.SecondFactorWebauthn:
+		return []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN}
+	case "":
+		// default to OTP
+		return []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_OTP}
 	default:
 		slog.WarnContext(context.Background(), "Found unknown second_factor setting", "second_factor", sf)
-		return "" // Unsure, say nothing.
+		return nil
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/authentication.go:382]: I think we should just validate `SecondFactor` in `CheckAndSetDefaults`.

File: api/types/authentication.go [hunk 10]
```diff
@@ -328,29 +328,12 @@ func (c *AuthPreferenceV2) GetSecondFactor() constants.SecondFactorType {
...
-	}
-}
-
 // SetSecondFactor sets the type of second factor.
 func (c *AuthPreferenceV2) SetSecondFactor(s constants.SecondFactorType) {
 	c.Spec.SecondFactor = s
 
 	// Set SecondFactors from SecondFactor.
-	c.Spec.SecondFactors = SecondFactorsFromLegacySecondFactor(c.Spec.SecondFactor)
+	c.Spec.SecondFactors = secondFactorsFromLegacySecondFactor(c.Spec.SecondFactor)
```
Comment (codingllama) [inline api/types/authentication.go:336]: Similar to other comments, should we set only the SecondFactor (singular) field here?

File: api/types/second_factor.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+	secondFactorTypeOTPString = "otp"
+	// secondFactorTypeWebauthnString is the string representation of SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN
+	secondFactorTypeWebauthnString = "webauthn"
+	// secondFactorTypeSSOString is the string representation of SecondFactorType_SECOND_FACTOR_TYPE_SSO
+	secondFactorTypeSSOString = "sso"
+)
+
+// ToString returns the user friendly string representation of the second factor type.
+func (s *SecondFactorType) ToString() string {
+	str, _ := s.encode()
```
Comment (codingllama) [inline api/types/second_factor.go:115]: Log the error?

File: api/types/second_factor.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+	case SecondFactorType_SECOND_FACTOR_TYPE_UNSPECIFIED:
+		return "", nil
+	case SecondFactorType_SECOND_FACTOR_TYPE_OTP:
+		return secondFactorTypeOTPString, nil
+	case SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN:
+		return secondFactorTypeWebauthnString, nil
+	case SecondFactorType_SECOND_FACTOR_TYPE_SSO:
+		return secondFactorTypeSSOString, nil
+	default:
+		return "", trace.BadParameter("SecondFactorType invalid value %v", *s)
```
Comment (codingllama) [inline api/types/second_factor.go:130]:
```suggestion
		return "", trace.BadParameter("invalid SecondFactorType value: %q", *s)
```

So we avoid capitalization at the start. Same for others.

- https://go.dev/wiki/CodeReviewComments#error-strings

File: api/types/second_factor.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+	case int64:
+		return trace.Wrap(s.setFromEnum(int32(v)))
+	case int:
+		return trace.Wrap(s.setFromEnum(int32(v)))
+	case float64:
+		return trace.Wrap(s.setFromEnum(int32(v)))
+	case float32:
+		return trace.Wrap(s.setFromEnum(int32(v)))
+	default:
+		return trace.BadParameter("RequireMFAType invalid type %T", val)
```
Comment (codingllama) [inline api/types/second_factor.go:160]:
```suggestion
		return trace.BadParameter("invalid RequireMFAType type: %T", val)
```

File: lib/config/fileconf.go (hunk 5 above)
Comment (Joerger) [reply lib/config/fileconf.go:1097]: This _is_ taking the cap as the user wrote it. But I will remove the warning log and update `CheckAndSetDefaults` to return `trace.BadParameter("must set either SecondFactor or SecondFactors, not both")` sot that this would return an error instead.

File: api/types/second_factor.go [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+	}
+	return nil
+}
+
+// setFromEnum sets the value from enum value as int32.
+func (r *SecondFactorType) setFromEnum(val int32) error {
+	if _, ok := SecondFactorType_name[val]; !ok {
+		return trace.BadParameter("invalid SecondFactorType mode %v", val)
```
Comment (codingllama) [inline api/types/second_factor.go:168]:
```suggestion
		return trace.BadParameter("invalid SecondFactorType enum: %v", val)
```

File: api/types/second_factor.go [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+	case float32:
+		return trace.Wrap(s.setFromEnum(int32(v)))
+	default:
+		return trace.BadParameter("RequireMFAType invalid type %T", val)
+	}
+	return nil
+}
+
+// setFromEnum sets the value from enum value as int32.
+func (r *SecondFactorType) setFromEnum(val int32) error {
```
Comment (codingllama) [inline api/types/second_factor.go:166]: Use `s` for the receiver, like the other methods?

File: lib/config/fileconf.go (hunk 5 above)
Comment (codingllama) [reply lib/config/fileconf.go:1097]: Right, sorry for the noise.

File: api/types/authentication.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply api/types/authentication.go]: Would erroring here make for a clearer error?

[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/authentication.go]: I don't think so

[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/authentication.go]: bf78d9e0232e748f6b323ac7895dbba85f3040ff

**Review by codingllama — APPROVED** [review]

File: api/types/authentication.go [hunk 16]
```diff
@@ -87,11 +87,12 @@ type AuthPreference interface {
 	IsSecondFactorEnabled() bool
 	// IsSecondFactorEnforced checks if second factor is enforced.
 	IsSecondFactorEnforced() bool
-	// IsSecondFactorTOTPAllowed checks if users are allowed to register TOTP devices.
+	// IsSecondFactorTOTPAllowed  checks if users can use TOTP as an MFA method.
```
Comment (codingllama) [inline api/types/authentication.go:90]:
```suggestion
	// IsSecondFactorTOTPAllowed checks if users can use TOTP as an MFA method.
```

File: api/types/authentication.go [hunk 17]
```diff
@@ -771,6 +775,16 @@ func (c *AuthPreferenceV2) CheckAndSetDefaults() error {
 		return trace.BadParameter("missing required Webauthn configuration for headless=true")
 	}
 
+	// Prevent accidental local lockout by disabling local second factor methods, (likely leaving only sso enabled).
```
Comment (codingllama) [inline api/types/authentication.go:778]: Could we get test coverage for this?

File: api/types/authentication.go [hunk 18]
```diff
@@ -722,7 +726,7 @@ func (c *AuthPreferenceV2) CheckAndSetDefaults() error {
 		c.Spec.SecondFactor = constants.SecondFactorWebauthn
 	case "":
 		// default to OTP if SecondFactors is also not set.
-		if len(c.Spec.SecondFactors) == 0 {
+		if c.Spec.SecondFactors == nil {
```
Comment (codingllama) [inline api/types/authentication.go:729]: The `len` check is usually better, `[]SecondFactorType{}` is not nil but it has zero length.

File: api/types/second_factor.go [hunk 19]
```diff
@@ -29,9 +29,10 @@ import (
 func legacySecondFactorFromSecondFactors(secondFactors []SecondFactorType) constants.SecondFactorType {
 	hasOTP := slices.Contains(secondFactors, SecondFactorType_SECOND_FACTOR_TYPE_OTP)
 	hasWebAuthn := slices.Contains(secondFactors, SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN)
+	hasSSO := slices.Contains(secondFactors, SecondFactorType_SECOND_FACTOR_TYPE_SSO)
 
 	switch {
-	case hasOTP && hasWebAuthn:
+	case (hasOTP && hasWebAuthn) || hasSSO:
```
Comment (codingllama) [inline api/types/second_factor.go:35]: This is an interesting semantical change. Up to this point SecondFactorOn meant that all configs had to be set, but now we could have only SSO second factor but have to call it "on" (because we lack a value to express it).

I think this is (mostly?) fine (we typically check for "otp available" or "webauthn config exists" as well), but we should update godocs on SecondFactorOn to mention the new corner-case.

We should also test this scenario in practice when we get to it. Maybe leave a TODO here as well?

File: api/types/authentication.go [hunk 20]
```diff
@@ -320,20 +320,20 @@ func (c *AuthPreferenceV2) SetType(s string) {
 
 // GetSecondFactor returns the type of second factor.
 func (c *AuthPreferenceV2) GetSecondFactor() constants.SecondFactorType {
-	if c.Spec.SecondFactor != "" {
-		return c.Spec.SecondFactor
+	// SecondFactors takes priority if set.
+	if c.Spec.SecondFactors != nil {
```
Comment (codingllama) [inline api/types/authentication.go:325]: Note that `!= nil` doesn't mean `len > 0`. A len check would be safer.

File: api/types/authentication.go (hunk 18 above)
Comment (Joerger) [reply api/types/authentication.go:729]: Oops, I really thought this wasn't the case. I guess only `var sfs []SecondFactorType` would evaluate to nil.

File: lib/config/configuration_test.go
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/config/configuration_test.go:1]: Is there any way for me to avoid this happening? Without gofumpt I get lint issues, but with it it makes unnecessary changes like this.

File: api/types/second_factor.go (hunk 19 above)
Comment (Joerger) [reply api/types/second_factor.go:35]: The only purpose of this is here - https://github.com/gravitational/teleport/pull/47426/commits/82eff9a725ead23ce6ec50cd67ff6d471499da22

If you'd like I can hold off on merging this PR until that one is also ready and approved

Comment (Joerger) [reply api/types/second_factor.go:35]: I wanted to set this to "on" since we are relying on the logic in some places. However this will be unnecessary as soon as https://github.com/gravitational/teleport/pull/47426/commits/82eff9a725ead23ce6ec50cd67ff6d471499da22 is merged, so I'd rather not overcomplicate things. If you'd like I can hold off on merging this until that PR is ready and approved.

File: api/types/second_factor.go
[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/second_factor.go:1]: Yes it should be covered well enough, especially once https://github.com/gravitational/teleport/pull/47426 to make `second_factors` used primarily.

File: api/types/authentication.go (hunk 17 above)
Comment (Joerger) [reply api/types/authentication.go]: 6ab4fdc5560b7900cb21ab497d5970d3d9fa19c9

File: api/types/second_factor.go (hunk 19 above)
Comment (codingllama) [reply api/types/second_factor.go]: I'm happy knowing that there is a follow up. I'd rather get this merged than wait.

File: api/types/authentication.go [hunk 21]
```diff
@@ -775,14 +775,11 @@ func (c *AuthPreferenceV2) CheckAndSetDefaults() error {
...
+	// Prevent local lockout by disabling local second factor methods.
+	hasSSO := c.IsSecondFactorSSOAllowed()
 	hasTOTP := c.IsSecondFactorTOTPAllowed()
-	if c.GetAllowLocalAuth() && !hasTOTP && !hasWebauthn {
-		switch c.Spec.SecondFactor {
-		case constants.SecondFactorOptional, constants.SecondFactorOff:
-		default:
-			return trace.BadParameter("missing a local second factor method for local users (otp, webauthn), either add a local second factor method or disable local auth")
-		}
+	if c.GetAllowLocalAuth() && hasSSO && !hasTOTP && !hasWebauthn {
```
Comment (codingllama) [inline api/types/authentication.go:781]: Why does `hasSSO` matter here? It looks like the test doesn't really care whether it's true of false.

File: api/types/authentication_test.go [hunk 22]
```diff
@@ -64,3 +68,144 @@ func TestEncodeDecodeRequireMFAType(t *testing.T) {
...
+		},
+		{
+			name: "OK OTP",
+			spec: AuthPreferenceSpecV2{
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_OTP,
+				},
+			},
+			assertAuthPref: func(t *testing.T, authPref AuthPreference) {
+				require.False(t, authPref.GetAllowPasswordless())
```
Comment (codingllama) [inline api/types/authentication_test.go:94]: 1. Assert GetSecondFactors / GetSecondFactor directly? I think we care more about the flag than the derived settings. Same for others.

2. nit: Use `assert` instead of `require`?

File: api/types/authentication_test.go [hunk 23]
```diff
@@ -64,3 +68,144 @@ func TestEncodeDecodeRequireMFAType(t *testing.T) {
...
+		{
+			name: "OK U2F config provided",
+			spec: AuthPreferenceSpecV2{
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN,
+				},
+				U2F: &U2F{
+					AppID: "https://localhost",
+				},
+			},
```
Comment (codingllama) [inline api/types/authentication_test.go:136]: No asserts?

File: api/types/authentication_test.go [hunk 24]
```diff
@@ -64,3 +68,144 @@ func TestEncodeDecodeRequireMFAType(t *testing.T) {
...
+		{
+			name: "NOK SecondFactor and SecondFactors both set",
+			spec: AuthPreferenceSpecV2{
+				SecondFactor: constants.SecondFactorWebauthn,
+				SecondFactors: []SecondFactorType{
+					SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN,
+				},
+			},
+			assertErr: func(t require.TestingT, err error, vals ...interface{}) {
+				require.ErrorAs(t, err, &trace.BadParameterError{})
```
Comment (codingllama) [inline api/types/authentication_test.go:147]: That's a bit strange, I would expected we needed to pass `**trace.BadParameterError` to errors.As, since it's `*trace.BadParameter` that implements `error`.

File: api/types/authentication_test.go [hunk 25]
```diff
@@ -64,3 +68,144 @@ func TestEncodeDecodeRequireMFAType(t *testing.T) {
 		})
 	}
 }
+
+func TestAuthPreference_CheckAndSetDefaults(t *testing.T) {
```
Comment (codingllama) [inline api/types/authentication_test.go:72]:
```suggestion
func TestNewAuthPreference_secondFactors(t *testing.T) {
```

File: api/types/second_factor.go [hunk 26]
```diff
@@ -29,10 +29,9 @@ import (
 func legacySecondFactorFromSecondFactors(secondFactors []SecondFactorType) constants.SecondFactorType {
 	hasOTP := slices.Contains(secondFactors, SecondFactorType_SECOND_FACTOR_TYPE_OTP)
 	hasWebAuthn := slices.Contains(secondFactors, SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN)
-	hasSSO := slices.Contains(secondFactors, SecondFactorType_SECOND_FACTOR_TYPE_SSO)
 
 	switch {
-	case (hasOTP && hasWebAuthn) || hasSSO:
+	case (hasOTP && hasWebAuthn):
```
Comment (codingllama) [inline api/types/second_factor.go:34]: nit:

```suggestion
	case hasOTP && hasWebAuthn:
```

File: api/types/second_factor_test.go [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+
+package types
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// TestEncodeDecodeSecondFactorType tests encoding/decoding of the SecondFactorType.
```
Comment (codingllama) [inline api/types/second_factor_test.go:25]:
```suggestion
```

This one is pretty clear already, imo.

File: api/constants/constants.go [hunk 28]
```diff
@@ -208,7 +208,8 @@ const (
 	// is required for all users.
 	SecondFactorWebauthn = SecondFactorType("webauthn")
 	// SecondFactorOn means that all 2FA protocols are supported and 2FA is
-	// required for all users.
+	// required for all users, or that SSO MFA is enabled which cannot be
+	// expressed through a different SecondFactorType.
```
Comment (codingllama) [inline api/constants/constants.go:212]: Is this still true given recent changes?

File: api/types/authentication_test.go (hunk 22 above)
Comment (Joerger) [reply api/types/authentication_test.go:94]: I don't really care about `SecondFactor` here and `GetSecondFactors` just returns the list we set above. My intention with these tests was to test these default settings like passwordless. I'll add a separate test for converting SecondFactor to SecondFactors.

File: api/types/authentication.go (hunk 21 above)
Comment (Joerger) [reply api/types/authentication.go:781]: The test does set `SecondFactors=[sso]`, but what we really care about is `if second factor is enforced && local second factors are disabled`. I'll add a `IsLocalMFAEnabled` helper to clarify things.

**Review by codingllama — APPROVED** [review]

File: lib/config/configuration_test.go
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/config/configuration_test.go:1]: I just needed to add this to my vscode settings:

```
"[go]": {
    "editor.formatOnSaveMode": "modifications",
},
```

**Review by rosstimothy — APPROVED** [review]

**Comment by Joerger** [comment]
> @Joerger did we ever got docs updates for this one?

No not yet, it's on my TODO list, along with SSO MFA docs
